raft/clock.go

```diff
@@ -1,280 +1,92 @@
+	return newClockChan(c.HeartbeatInterval)
 }
 
-// clockTimers represents a list of sortable timers.
-type clockTimers []clockTimer
-
-func (a clockTimers) Len() int           { return len(a) }
-func (a clockTimers) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a clockTimers) Less(i, j int) bool { return a[i].Next().Before(a[j].Next()) }
+// AfterReconnectTimeout returns a channel that fires after the reconnection timeout.
+func (c *Clock) AfterReconnectTimeout() <-chan chan struct{} { return newClockChan(c.ReconnectTimeout) }
 
-// Timer represents a single event.
-// The current time will be sent on C, unless the timer was created by AfterFunc.
-type Timer struct {
-	C     <-chan time.Time
-	c     chan time.Time
-	timer *time.Timer // realtime impl, if set
-	next  time.Time   // next tick time
-	mock  *mockClock  // mock clock, if set
-	fn    func()      // AfterFunc function, if set
-}
+// AfterWaitInterval returns a channel that fires after the wait interval.
+func (c *Clock) AfterWaitInterval() <-chan chan struct{} { return newClockChan(c.WaitInterval) }
 
-// Stop turns off the ticker.
-func (t *Timer) Stop() {
-	if t.timer != nil {
-		t.timer.Stop()
-	} else {
-		t.mock.removeClockTimer((*internalTimer)(t))
-	}
+// HeartbeatTicker returns a Ticker that ticks every heartbeat.
+func (c *Clock) HeartbeatTicker() *Ticker {
+	t := time.NewTicker(c.HeartbeatInterval)
+	return &Ticker{C: t.C, ticker: t}
 }
 
-type internalTimer Timer
-
-func (t *internalTimer) Next() time.Time { return t.next }
-func (t *internalTimer) Tick(now time.Time) {
-	if t.fn != nil {
-		t.fn()
-	} else {
-		t.c <- now
-	}
-	t.mock.removeClockTimer((*internalTimer)(t))
-	gosched()
-}
+// Now returns the current wall clock time.
+func (c *Clock) Now() time.Time { return time.Now() }
 
 // Ticker holds a channel that receives "ticks" at regular intervals.
 type Ticker struct {
 	C      <-chan time.Time
-	c      chan time.Time
-	ticker *time.Ticker  // realtime impl, if set
-	next   time.Time     // next tick time
-	mock   *mockClock    // mock clock, if set
-	d      time.Duration // time between ticks
+	ticker *time.Ticker // realtime impl, if set
 }
 
 // Stop turns off the ticker.
 func (t *Ticker) Stop() {
 	if t.ticker != nil {
 		t.ticker.Stop()
-	} else {
-		t.mock.removeClockTimer((*internalTicker)(t))
 	}
 }
 
-type internalTicker Ticker
-
-func (t *internalTicker) Next() time.Time { return t.next }
-func (t *internalTicker) Tick(now time.Time) {
-	select {
-	case t.c <- now:
-	case <-time.After(goschedTimeout):
-	}
-	t.next = now.Add(t.d)
-	gosched()
-}
-
-// goschedTimeout is the amount of wall time to sleep during goroutine scheduling.
-var goschedTimeout = 1 * time.Millisecond
-
-// Sleep momentarily so that other goroutines can process.
-func gosched() {
-	time.Sleep(goschedTimeout)
+// newClockChan returns a channel that sends a channel after a given duration.
+// The channel being sent can be used to notify the sender when an action is done.
```

Comment:
This comment is ever-so slight ambiguous. When I read it first I read "sent" as synonymous with "returned", which is sloppy on my part. Still, the next guy might make the same mistake. Perhaps make it crystal clear?

`The channel being, sent over the channel that is returned, can be used to notify the sender when an action is done.`